### PR TITLE
Split traces properly, so that every expression has a unique trace

### PR DESCRIPTION
**Pull request by hjorthjort**
At some points in `source_to_mod` and `exh_to_pat` we created many expressions with the same trace. This pull request fixes that.

### CODE REVIEW

File: compiler/backend/exh_to_patScript.sml [hunk 1]
```diff
@@ -3,23 +3,25 @@ open backend_commonTheory
...
 val sIf_def = Define `
-  sIf tra e1 e2 e3 =
+  sIf t e1 e2 e3 =
   if e2 = Bool T ∧ e3 = Bool F
     then e1
   else
     (dtcase e1 of
      | Con _ t [] => if t = true_tag then e2 else e3
-     | _ => If tra e1 e2 e3)`;
+     | _ => If t e1 e2 e3)`;
```
Comment (Holmus) [inline compiler/backend/exh_to_patScript.sml]: A bit confusing to use the name t both for the boolean and the trace

File: compiler/backend/source_to_modScript.sml [hunk 2]
```diff
@@ -60,23 +67,25 @@ val compile_exp_def = tDefine"compile_exp"`
     | SOME x => x)
   ∧
   (compile_exp t env (Fun x e) =
-    Fun t x (compile_exp t (nsBind x (Var_local t x) env) e))
+    let (t1, t2) = (mk_cons t 1, mk_cons t 2) in
+      Fun t1 x (compile_exp t (nsBind x (Var_local t1 x) env) e))
```
Comment (Holmus) [inline compiler/backend/source_to_modScript.sml]: Assuming Var_local should have t2, otherwise this achieves no change

**Review by Holmus — CHANGES_REQUESTED** [review]

**Comment by myreen** [comment]
The `pat_tups` function produces traces `t`, `mk_cons ... t`, `mk_cons ... (mk_cons ... t)`, `mk_cons ... (mk_cons ... (mk_cons ... t))`. To me this is wrong. A trace should either pass over unchanged or the traces in the next language should all be of the form `mk_cons ... t`, if `t` is the original trace. I don't think you want traces to be prefixes of other traces in the same language.

File: compiler/backend/source_to_modScript.sml (hunk 2 above)
Comment (hjorthjort) [reply compiler/backend/source_to_modScript.sml]: Yes, exactly. Good catch.

File: compiler/backend/exh_to_patScript.sml (hunk 1 above)
Comment (hjorthjort) [reply compiler/backend/exh_to_patScript.sml]: Completely right. Changing back.

**Comment by hjorthjort** [comment]
Completely right, Magnus. It was supposed to just add a number with `mk_cons t n`, where `n` would shrink for every iteration. But by overshadowing `t` in the `let` part causes the effect you described. This is fixed now by creating `t'`

```
val pat_tups_def = Define`
  (pat_tups t [] = [])
  /\
  (pat_tups t (x::xs) =
    let t' = mk_cons t ((LENGTH xs) + 1) in
      (x, Var_local t' x)::pat_tups t xs)`;
```

**Pull request merged**